#### Paphos Weather

This page shows various metrics for the weather in Paphos, Cyprus.

Weather data is purchased from Openweather Marketplace [here](https://home.openweathermap.org/marketplace/my_orders)



In [2]:

from datetime import date
from datetime import datetime
import time
import pandas as pd


def toDate(epoch_time):
    return datetime.fromtimestamp(epoch_time)
  


In [1]:
import pandas as pd

df=pd.read_json('/Users/walkerrowe/Documents/weather/HypatiaAcademy/assignment/pathosWeather.json')

In [3]:
toDate(946684800)

datetime.datetime(1999, 12, 31, 19, 0)

In [4]:
df['dt'].map(toDate)

0        1999-12-31 19:00:00
1        1999-12-31 20:00:00
2        1999-12-31 21:00:00
3        1999-12-31 22:00:00
4        1999-12-31 23:00:00
                 ...        
181459   2020-09-12 15:00:00
181460   2020-09-12 16:00:00
181461   2020-09-12 17:00:00
181462   2020-09-12 18:00:00
181463   2020-09-12 19:00:00
Name: dt, Length: 181464, dtype: datetime64[ns]

In [5]:
df.columns

Index(['city_name', 'lat', 'lon', 'main', 'wind', 'clouds', 'weather', 'dt',
       'dt_iso', 'timezone', 'rain'],
      dtype='object')

In [6]:
df['main'][0]

{'temp': 55.35,
 'temp_min': 51.8,
 'temp_max': 65.53,
 'feels_like': 49.44,
 'pressure': 1016,
 'humidity': 73}

We have to convert the epoch time to ISO Date, which is something we can read.

The map function runs a function over a series.  A series means a dataframe with one column.  When we write df['new column'] = df['some column'].map(somefunction) it adds the new columns to the dataframe.


In [17]:
df['isoDate']=df['dt'].map(toDate)

In [19]:
df.columns

Index(['city_name', 'lat', 'lon', 'main', 'wind', 'clouds', 'weather', 'dt',
       'dt_iso', 'timezone', 'rain', 'isoDate'],
      dtype='object')

In [18]:
df['isoDate']

0        1999-12-31 19:00:00
1        1999-12-31 20:00:00
2        1999-12-31 21:00:00
3        1999-12-31 22:00:00
4        1999-12-31 23:00:00
                 ...        
181459   2020-09-12 15:00:00
181460   2020-09-12 16:00:00
181461   2020-09-12 17:00:00
181462   2020-09-12 18:00:00
181463   2020-09-12 19:00:00
Name: isoDate, Length: 181464, dtype: datetime64[ns]

We want to caculate the average temperature by date usign the mean() function.  But the problem is that the date is an object inside the **main** object.  See below how to fix that.

In [9]:
df.index = pd.to_datetime(df['isoDate'])
df.resample('d').mean()

lat        lon            dt  timezone
isoDate                                                 
1999-12-31  34.753637  32.406951  9.466920e+08    7200.0
2000-01-01  34.753637  32.406951  9.467442e+08    7200.0
2000-01-02  34.753637  32.406951  9.468306e+08    7200.0
2000-01-03  34.753637  32.406951  9.469170e+08    7200.0
2000-01-04  34.753637  32.406951  9.470034e+08    7200.0
...               ...        ...           ...       ...
2020-09-08  34.753637  32.406951  1.599579e+09   10800.0
2020-09-09  34.753637  32.406951  1.599665e+09   10800.0
2020-09-10  34.753637  32.406951  1.599752e+09   10800.0
2020-09-11  34.753637  32.406951  1.599838e+09   10800.0
2020-09-12  34.753637  32.406951  1.599917e+09   10800.0

[7562 rows x 4 columns]

The series df['main'] is a JSON object.  Write a function to split it into it's seperate fields.  Use **series.map(function)** method like above.

{'temp': 55.35,
 'temp_min': 51.8,
 'temp_max': 65.53,
 'feels_like': 49.44,
 'pressure': 1016,
 'humidity': 73}
 

In [16]:
df.columns

Index(['city_name', 'lat', 'lon', 'main', 'wind', 'clouds', 'weather', 'dt',
       'dt_iso', 'timezone', 'rain'],
      dtype='object')

In [11]:
df.head(1)

city_name        lat        lon  \
isoDate                                                    
1999-12-31 19:00:00  Paphos Castle  34.753637  32.406951   

                                                                  main  \
isoDate                                                                  
1999-12-31 19:00:00  {'temp': 55.35, 'temp_min': 51.8, 'temp_max': ...   

                                           wind      clouds  \
isoDate                                                       
1999-12-31 19:00:00  {'speed': 9.17, 'deg': 20}  {'all': 1}   

                                                               weather  \
isoDate                                                                  
1999-12-31 19:00:00  [{'id': 800, 'main': 'Clear', 'description': '...   

                            dt                         dt_iso  timezone rain  \
isoDate                                                                        
1999-12-31 19:00:00  946684800  2000-01-01 00:00:00 +0000 UTC      7200  NaN   

                                isoDate  
isoDate                                  
1999-12-31 19:00:00 1999-12-31 19:00:00